v2/writer.go

```diff
@@ -79,6 +80,99 @@ func WrapV1(src io.ReadSeeker, dst io.Writer) error {
 	return nil
 }
 
+// ExtractV1File takes a CAR file and extracts its CARv1 data payload, unmodified.
+// The resulting CARv1 file will not include any data payload padding that may be present in the
+// CARv2 srcPath.
+// The srcPath is assumed to exist, and the destination path is created if not exist.
+// Note that the destination path might still be created even if an error
+// occurred.
+// If srcPath and dstPath are the same, then the dstPath is converted, in-place, to CARv1.
+//
+// The srcPath may represent a CARv1 or a CARv2 file.
+// If the srcPath represents a CARv1 file, the entire content is copied to dstPath unchanged
+// effectively copying the source file.
+// In a case where srcPath and dstPath are the same and represent a CARv1, this function returns
+// immediately without writing any bytes.
+func ExtractV1File(srcPath, dstPath string) error {
+	src, err := os.Open(srcPath)
+	if err != nil {
+		return err
+	}
+	defer src.Close()
+
+	dst, err := os.OpenFile(dstPath, os.O_CREATE|os.O_WRONLY, 0o666)
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+
+	// Detect CAR version.
+	version, err := ReadVersion(src)
+	if err != nil {
+		return err
+	}
+	switch version {
+	case 1:
+		// If source is a CARv1 and source and destination paths are the same then we are done.
+		if srcPath == dstPath {
+			return nil
+		}
+		// Seek back to the beginning, since pragma was read earlier to determine version.
+		if _, err := src.Seek(0, io.SeekStart); err != nil {
+			return err
+		}
+		// Copy all of src into dst.
+		if _, err := io.Copy(dst, src); err != nil {
+			return err
+		}
+	case 2:
+		// Read CARv2 header to locate data payload.
+		v2h := &Header{}
+		if _, err := v2h.ReadFrom(src); err != nil {
+			return err
+		}
+
+		// TODO consider extracting this into Header.Validate since it is also implemented in BlockReader.
+		// Validate header
+		dataOffset := int64(v2h.DataOffset)
+		if dataOffset < PragmaSize+HeaderSize {
+			return fmt.Errorf("invalid data payload offset: %v", dataOffset)
+		}
+		dataSize := int64(v2h.DataSize)
+		if dataSize <= 0 {
+			return fmt.Errorf("invalid data payload size: %v", dataSize)
+		}
+
+		// Seek to the point where the data payload starts
+		if _, err := src.Seek(dataOffset, io.SeekStart); err != nil {
+			return err
+		}
+
+		// Copy data payload over, expecting to write exactly the right number of bytes
+		written, err := io.CopyN(dst, src, dataSize)
```

Comment:
Are we sure that this will delegate to kernel? If so, can we add a comment here explaining what we expect to happen?

Reply:
Comments added. thank you for pointing this out.

 Re "Are we sure":
I have implemented two benchmarks: one uses `ExtractV1File`, and the other user `Reader` API. The expectation is that the former is more efficient. The point is to measure this.